src/Interpreters/OptimizeShardingKeyRewriteInVisitor.cpp

```diff
@@ -191,6 +243,61 @@ class OptimizeShardingKeyRewriteIn : public InDepthQueryTreeVisitorWithContext<O
         }
     }
 
+    void rewriteGlobalInPredicate(FunctionNode * function_node)
+    {
+        auto & arguments = function_node->getArguments().getNodes();
+
+        const auto & required_columns = data.sharding_key_expr->getRequiredColumns();
+        if (required_columns.size() != 1)
+            return;
+
+        const auto & required_column_name = required_columns.front();
+
+        auto * column_in = arguments[0]->as<ColumnNode>();
+        if (!column_in || column_in->getColumnName() != required_column_name)
+            return;
+
+        const auto & source_table = arguments[1];
+        if (!source_table->as<TableNode>())
+            return;
+
+        const auto * sharding_key_expr_node = data.sharding_key_expr->getActionsDAG().tryFindInOutputs(data.sharding_key_column_name);
+        if (!sharding_key_expr_node)
+            return;
+
+        const auto & subquery = buildSubqueryToReadColumnsFromTableExpression(source_table, getContext());
+        const auto & subquery_column = subquery->as<QueryNode &>().getProjection().getNodes().front();
+
+        std::map<std::string, QueryTreeNodePtr> allowed_inputs;
+        allowed_inputs[required_column_name] = subquery_column;
+
+        QueryTreeNodePtr sharding_key_filter;
+        try
+        {
+            sharding_key_filter = actionsDAGNodeToQueryTreeNode(*sharding_key_expr_node, allowed_inputs);
+        }
+        catch (const Exception & e)
+        {
+            if (e.code() == ErrorCodes::NOT_IMPLEMENTED)
+                return;
+
+            throw;
+        }
```

Comment:
Better to build a query tree for the sharding key in the `StorageDistributed` constructor using AST.

Reply:
Will do, marking as draft for now.

Reply:
Since storing raw query tree node (without identifiers resolved) in `StorageDistribute` doesn't seem to make much sense, and even in case the identifiers are resolved, we need to have the exact table expression, used while query analysis, for column replacement, it seems easier to pass AST as an argument and localize building query tree, query analysis and column replacement for sharding key in this rewrite visitor, instead of passing both query tree and table expression, that has been used during analysis.

The updated solution with passing `ASTPtr` also doesn't require conversion of `ActionsDAG` to query tree.